internal/expr/parser.go

```diff
@@ -368,64 +439,64 @@ func (p *Parser) parseIdentifier() (string, bool) {
 // parseColumn parses a column made up of name bytes, optionally dot-prefixed by
 // its table name.
 // parseColumn returns an error so that it can be used with parseList.
-func (p *Parser) parseColumn() (fullName, bool, error) {
+func (p *Parser) parseColumn() (columnName, bool, error) {
 	cp := p.save()
 
 	if id, ok := p.parseIdentifierAsterisk(); ok {
 		if id != "*" && p.skipByte('.') {
 			if idCol, ok := p.parseIdentifierAsterisk(); ok {
-				return fullName{prefix: id, name: idCol}, true, nil
+				return columnName{table: id, name: idCol}, true, nil
 			}
 		} else {
 			// A column name specified without a table prefix should be in name.
-			return fullName{name: id}, true, nil
+			return columnName{name: id}, true, nil
 		}
 	}
 
 	cp.restore()
-	return fullName{}, false, nil
+	return columnName{}, false, nil
 }
 
-func (p *Parser) parseTargetType() (fullName, bool, error) {
+func (p *Parser) parseTargetType() (typeName, bool, error) {
 	if p.skipByte('&') {
-		return p.parseGoFullName()
+		return p.parseTypeName()
 	}
 
-	return fullName{}, false, nil
+	return typeName{}, false, nil
 }
 
-// parseGoFullName parses a Go type name qualified by a tag name (or asterisk)
-// of the form "&TypeName.col_name".
-func (p *Parser) parseGoFullName() (fullName, bool, error) {
+// parseTypeName parses a Go type name qualified by a tag name (or asterisk)
+// of the form "TypeName.col_name".
+func (p *Parser) parseTypeName() (typeName, bool, error) {
 	cp := p.save()
 
 	if id, ok := p.parseIdentifier(); ok {
 		if !p.skipByte('.') {
-			return fullName{}, false, fmt.Errorf("column %d: unqualified type, expected %s.* or %s.<db tag>", p.pos, id, id)
+			return typeName{}, false, fmt.Errorf("column %d: unqualified type, expected %s.* or %s.<db tag>", p.pos, id, id)
 		}
 
 		idField, ok := p.parseIdentifierAsterisk()
 		if !ok {
-			return fullName{}, false, fmt.Errorf("column %d: invalid identifier suffix following %q", p.pos, id)
+			return typeName{}, false, fmt.Errorf("column %d: invalid identifier suffix following %q", p.pos, id)
 		}
-		return fullName{id, idField}, true, nil
+		return typeName{name: id, member: idField}, true, nil
 	}
 
 	cp.restore()
-	return fullName{}, false, nil
+	return typeName{}, false, nil
 }
 
-// parseList takes a parsing function that returns a fullName and parses a
+// parseList takes a parsing function that returns a T and parses a
 // bracketed, comma seperated, list.
-func (p *Parser) parseList(parseFn func(p *Parser) (fullName, bool, error)) ([]fullName, bool, error) {
+func parseList[T any](p *Parser, parseFn func(p *Parser) (T, bool, error)) ([]T, bool, error) {
```

Comment:
So what's the actual goal of the change?  On a quick sight, it seems to now be using generics while before we didn't (changes min version of Go), and is splitting code that was common into independent functions so that they can handle the different types.

Curiously, this is how things were originally, until we decided to unify to make logic common. So has something changed again to make it feel like it should be split once more?

Reply:
The goal of the change is to differentiate between column names and type names in the type system. Before this change we were using `fullName` for both. The problem is that it was not clear where a function or some part of the code was expecting one or the other. Now, it is clear just by looking at the type.

The downsides, as you correctly state, is that this function has to be generic because we want it to work with both types, and that there is additional code duplication (i.e. `starCountTypes` and `starCountColumns`). For the latter, even though the code *seems* duplicated because several functions are similar, in reality the types are semantically different which leads to some frustration when mixing them just because the implementation can be shared.